aldryn_faq/tests/frontend/integration/specs/spec.faq.crud.js

```diff
@@ -35,4 +35,79 @@ describe('Aldryn FAQ tests: ', function () {
         });
     });
 
+    it('creates a new test page', function () {
+        // click the example.com link in the top menu
+        faqPage.userMenus.first().click().then(function () {
+            // wait for top menu dropdown options to appear
+            browser.wait(function () {
+                return browser.isElementPresent(faqPage.userMenuDropdown);
+            }, faqPage.mainElementsWaitTime);
+
+            faqPage.administrationOptions.first().click();
+        }).then(function () {
+            // wait for modal iframe to appear
+            browser.wait(function () {
+                return browser.isElementPresent(faqPage.sideMenuIframe);
+            }, faqPage.iframeWaitTime);
+
+            // switch to sidebar menu iframe
+            browser.switchTo().frame(browser.findElement(
+                By.css('.cms_sideframe-frame iframe')));
+
+            browser.wait(function () {
+                return browser.isElementPresent(faqPage.pagesLink);
+            }, faqPage.mainElementsWaitTime);
+
+            faqPage.pagesLink.click();
+
+            // check if the page already exists and return the status
+            return faqPage.addPageLink.isPresent();
+        }).then(function (present) {
+            if (present === true) {
+                // page is absent - create new page
+                browser.wait(function () {
+                    return browser.isElementPresent(faqPage.addPageLink);
+                }, faqPage.mainElementsWaitTime);
+
+                faqPage.addPageLink.click();
+
+                browser.wait(function () {
+                    return browser.isElementPresent(faqPage.titleInput);
+                }, faqPage.mainElementsWaitTime);
+
+                faqPage.titleInput.sendKeys('Test').then(function () {
```

Comment:
yes, in general there's a possibility to optimise the code, but i don't know if the promise implementation in protractor is Promises/A+ compliant (i'm gonna assume it is) then you can rely on the behaviour `throw an error -> promise is rejected` so you could rewrite bunch of code like this:

```js
p.then(function (present) {
    if (!present) {
       throw new Error('Element is not present :(');
    }
}).then(function () {
    // here error wasn't thrown so promise is fullfilled
    browser.wait(...);
    // ...

    // return a promise, don't nest `then` calls
    return faqPage.titleInput.sendKeys('Test');
}).then(function () {
    faqPage.saveButton.click();

    return faqPage.slugErrorNotification.isPresent();
}).then(function (present) {
    if (present) {
        throw new Error('Element is present :(');
    }
}).then(function () {
     browser.wait(function () {
        return browser.isElementPresent(faqPage.editPageLink);
    }, faqPage.mainElementsWaitTime);

    // etc etc
});